core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.function.Consumer;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * A dynamic {@link EndpointGroup}. List of {@link Endpoint} will be update dynamically.
+ */
+public abstract class DynamicEndpointGroup implements EndpointGroup {
+    private final Set<Consumer<Collection<Endpoint>>> updateListeners = new CopyOnWriteArraySet<>();
+
+    /**
+     * Notify the {@link Endpoint}s changes to listeners.
+     */
+    protected final void updateEndpoints(List<Endpoint> endpoints) {
```

Comment:
I wonder if it's better making `DynamicEndpointGroup` keep the list of `Endpoint`s, provide the methods that manipulate the list and make sure the notification is done automatically. i.e.

```java
public abstract class DyncmicEndpointGroup ... {
    private volatile List<Endpoint> endpoints = ...;
    protected final addEndpoint(Endpoint e) {
        ...
        newEndpointsBuilder.add(e);
        this.endpoints = newEndpointsBuilder.build();
        notifyListeners();
    }

    protected final removeEndpoint(Endpoint e) { ... notifyListener(); }
    protected final setEndpoints(Iterable<? extends Endpoint> endpoints) {
        this.endpoints = ImmutableList.copyOf(endpoints);
        notifyListeners();
    }

    @Override
    public final List<Endpoint> endpoints() {
        return endpoints;
    }
}
```